Implemented SeriesBase and FourierSeries class

leosartaj opened the pull request:
This PR brings in the following
- A `SeriesBase` class.
- `FourierSeries` class, that inherits from `SeriesBase`
### FourierSeries
- `FourierSeries` uses sequences. See #9435.
- `FourierSeries` is an implementation of fourier sine/cosine series.
- I have tried my best to speed up the computation. However, it's still slow. Probably due to 
  the required integration. To overcome, this I have added various operations(scale, shift, etc) that can be used for computing `FourierSeries` of other functions, if a `FourierSeries` is known.

closes #9077

ping @jcrist @flacjacket

leosartaj committed:
first version of SeriesBase class

leosartaj committed:
Implemented FourierSeries

Represents fourier sine/cosine series

Examples
========

```
>>> from sympy import FourierSeries, pi, cos
>>> from sympy.abc import x

>>> s = FourierSeries(x**2, (x, -pi, pi))
>>> s.as_series(n=3)
-4*cos(x) + cos(2*x) + pi**2/3
```

Shifting

```
>>> s.shift(1)
FourierSeries(x**2 + 1, (x, -pi, pi))
>>> s.shiftx(1)
FourierSeries((x + 1)**2, (x, -pi, pi))
```

Scaling

```
>>> s.scale(2)
FourierSeries(2*x**2, (x, -pi, pi))
>>> s.scalex(2)
FourierSeries(4*x**2, (x, -pi, pi))
```

Differentiating

```
>>> s.diff(x)
FourierSeries(2*x, (x, -pi, pi))
```

Notes
=====

Computing a fourier series can be slow
due to the integration required in computing
an, bn.

It is faster to compute fourier series of a function
by using shifting and scaling on an already
computed fourier series rather than computing
again.

eg. If Fourier series of x**2 is known
fourier series of x**2 - 1 can be found by shifting by 1

References
=========

.. [1] mathworld.wolfram.com/FourierSeries.html

flacjacket commented on sympy/series/fourier.py:
```diff
@@ -0,0 +1,391 @@
+            a0 = an[0] / 2
+
+
+        obj = Expr.__new__(cls, func, limits)
```
You also need to pass `an` and `bn` into the args. SymPy rebuilds objects (e.g. on a `.subs()`) from the args, whatever you pass to the args should be able to re-create the same object.

That said, I'm not sure how `an` and `bn` are supposed to work as kwargs, is it supposed to override the computation? should you be able to pass `an` and `bn` instead of `func`? it seems to not have any effect if I pass them.

leosartaj replied:
> You also need to pass an and bn into the args. SymPy rebuilds objects (e.g. on a .subs()) from the args, whatever you pass to the args should be able to re-create the same object.

I am not sure. Suppose an object with an and bn passed is subbed, in that case subs will pass an and bn again. Overriding the computation, hence giving wrong results. I will look into subs, and try to make it behave as expected.

> That said, I'm not sure how an and bn are supposed to work as kwargs, is it supposed to override the computation? should you be able to pass an and bn instead of func?

Well, if both an and bn are passed, no computation takes place. This is for speeding up the object creation. Function needs to be passed. They are just for saving some time. That said, they should not used by the end user. I added them to use in scale, shift, etc. I will add a note.

> it seems to not have any effect if I pass them.

Eg.
```
>>> FourierSeries(x**2) # takes time
>>> FourierSeries(x**2, an=sequence(cos(x)), bn=sequence(sin(x))) # instantaneous 
```

flacjacket commented on sympy/series/fourier.py:
```diff
@@ -0,0 +1,391 @@
+
+        i, j = 0, 0
+        terms = []
+        while(i < n):
```
Don't need the parens, `while i < n:`

Maybe do `while len(terms) < n)`?

leosartaj replied:
:+1:

jcrist replied:
Ridiculous itertools implementation. 

```
from itertools import ifilter, count
gen = ifilter(lambda a: a is not S.Zero, (self.term(j) for j in count(0)))
return [v for v, _ in zip(gen, range(n))]
```

leosartaj replied:
That is very compact. Thanks!

flacjacket commented on sympy/series/fourier.py:
```diff
@@ -0,0 +1,391 @@
+
+        return FourierSeries(sfunc, self.args[1], a0=a0, an=an, bn=bn)
+
+    def shiftx(self, s):
```
I'm not sure how necessary this is, can't a user just do a `.subs()` on the FourierSeries object?

jcrist replied:
I agree with Sean on this, I'm not sure if this method (or `scalex`) is needed.

flacjacket commented on sympy/series/series_class.py:
```diff
@@ -0,0 +1,108 @@
+
+        TODO
+        """
+        fsyms = set().union(*[a.free_symbols for a in self.args])
```
Ideally you wouldn't do the list comprehension before creating the sets, but my bit of playing with it wasn't able to get a workable solution.

jcrist replied:
`set(i for i in a.free_symbols for a in self.args)`

jcrist replied:
If you really wanted to be condensed, this whole block could be:

`return set(j for i in self.args for j in i.free_symbols if j not in self.variables)`

flacjacket commented on sympy/series/series_class.py:
```diff
@@ -0,0 +1,108 @@
+
+    def __iter__(self):
+        i = 0
+        while(i < self.length):
```
As above

jcrist replied:
While loops and manual integrating are antipatterns in python. A for loop would be better:

```
for i in range(self.length):
    pt = self._ith_point(i)
    yield self.term(pt)
```

Or better yet:
```
def __iter__(self):
    return (self.term(self._ith_pt(i)) for i in range(self.length))

leosartaj replied:
length here is ``oo`` and ``range(oo)`` does not work.

leosartaj commented:
if you could have a look @jcrist

jcrist commented on sympy/series/series_class.py:
```diff
@@ -0,0 +1,108 @@
+        Examples
+        ========
+
+        TODO
```
What's this todo about?

jcrist commented on sympy/series/series_class.py:
```diff
@@ -0,0 +1,108 @@
+        else:
+            initial = self.start
+
+        if self.start is S.NegativeInfinity:
```
These two if statements can be merged into one.

jcrist commented on sympy/series/series_class.py:
```diff
@@ -0,0 +1,108 @@
+        """
+        Returns the i'th point of a series
+        If start point is negative infinity, point is returned from the end.
+        Assumes the first point to be indexed zero.
```
I'm not sure if `start` and `stop` make sense this way, if this is how the code interprets them. If you always start at `0`, why not enforce `start` to be `0` then, and allow `stop` to be either `NegativeInfinity` or `PositiveInfinity`?

leosartaj replied:
This could also be a base to ``LaurentSeries`` maybe. So, start could be say -5 in that case. Interval -> (-5, oo). So, I just kept it this way.

jcrist commented on sympy/series/fourier.py:
```diff
@@ -0,0 +1,391 @@
+
+
+        obj = Expr.__new__(cls, func, limits)
+        obj.a0, obj.an, obj.bn = a0, an, bn
```
SymPy objects shouldn't stores things anywhere but in `args`. If `a0, an, bn` are part of the internal representation, then they should be in `args`. Note that `args` doesn't have to be what's passed into the user facing interface, just how it's internally represented. However, `object == object.func(*object.args)` must also be true (`func` is `__new__` by default). In this case, I'd make these part of args, and have `__new__` somehow determine (without kwargs) whether what was passed in was `a0, an, bn` or `func, limits`. A simple `len(args)` would probably suffice.

jcrist commented:
General comments:

1. `SeriesBase` looks fine as is. I'm sure as you're implementing more series things, changes will happen to the base class. I'm not overly concerned about it.

2. Think about the internal representation of `FourierSeries`. In SymPy, ideally all internal representation is stored in `args`. Some classes don't follow this for various reasons, but overall this is an idiom that should be followed. At the same time, you must ensure that `expr == expr.func(*expr.args)` for all cases. I'm not sure what the correct solution to this is for `FourierSeries`, but it's this is a blocking issue until it's resolved.

leosartaj commented:
Thanks for reviewing this.

Some issues I am facing:

* Printing - Will this Work?

```
>>> FourierSeries(x)
2*sin(x) - sin(2*x) + 2*sin(3*x) / 3 + ...
```
I am not really sure if ``Order`` can be used here. As terms are sin / cos terms only.

* Subs/scalex/shiftx - I primarily added ``shiftx`` and ``scalex`` to save some computation time, like changing function from ``x**2`` to ``4*(x + 1)**2`` can be done using ``scalex`` and ``shiftx``, it will be fast. But this can/should also be done when using ``subs``. I am not really how to do this. I tried implementing ``_eval_subs``, but I think it will be tricky to allow it to identify and use ``scale`` and ``shift`` functions or I could keep the default behaviour, where ``FourierSeries`` is again computed. All in all, I think ``shiftx`` and ``scalex`` should be kept.

Thoughts?
@jcrist @flacjacket

leosartaj commented:
Should ``as_series`` work the way it does? Or is their a better way of representing truncated ``FourierSeries`` in your opinion?

leosartaj committed:
api changes to FourierSeries

leosartaj committed:
removed itertools from as_series in favour of for loop

leosartaj committed:
implemented fourier_series function

jcrist commented:
Sorry for the delay here.

>  Printing - Will this Work?

Printing looks good to me. I'm fairly certain you can't use order here. Someone with a better math background could probably answer this better (@flacjacket). The `...` looks good to me though.

> All in all, I think shiftx and scalex should be kept.

Thinking about it some more, `subs(x, ...)` would have to avoid performing on the limit, but for symbolic limits, you'd need to traverse the limit for subs involving terms inside the limit, otherwise you'd get incorrect results. So this is fine, sorry for sending you down a wrong path earlier.

> Should `as_series` work the way it does? Or is their a better way of representing truncated FourierSeries in your opinion?

The way it works is fine, but I would change the name, as it doesn't return a `Series` object. Something like `truncate(n)`, `as_expr(n)`, or `as_add(n)` may be better. I like `truncate` the best.

leosartaj commented:
``truncate`` looks good to me.

leosartaj committed:
Added printing support, renamed as_series to truncate

leosartaj commented:
I have added printing support and renamed ``as_series`` to ``truncate``. ``FourierSeries`` is now ready for a final review.
Ping @jcrist @flacjacket

leosartaj committed:
more tests added, coverage 100%

leosartaj commented:
All the computation is now done inside ``fourier_series``. ``FourierSeries`` class is only for representational purpose. I have further cleaned up args.
 
Ping @jcrist @flacjacket

leosartaj committed:
No computation inside FourierSeries class

leosartaj committed:
some code cleanup

leosartaj commented:
Anything needs to be done?
ping @jcrist

flacjacket commented on sympy/printing/tests/test_latex.py:
```diff
@@ -538,6 +538,11 @@ def test_latex_sequences():
 
 
+def test_latex_FourierSeries():
+    latex_str = r'2 \sin{\left (x \right )} - \sin{\left (2 x \right )} + \frac{2}{3} \sin{\left (3 x \right )}\ldots'
```
The normal print methods have `+ ...` at the end rather than `...`, the latex should probably do the same

leosartaj committed:
fix latex printing of FourierSeries

leosartaj commented:
I fixed the latex printing. @flacjacket

flacjacket commented:
Thanks @leosartaj is there anything else with this? Is it ok to merge as is? I know #9572 is on top of this, I don't want to mess up your other branches.

leosartaj commented:
I don't think there is any problem with ``FourierSeries`` as of now. For ``SeriesBase``, if there are any changes they can be brought in #9572. I think this is mergeable.
Thanks for reviewing.

flacjacket commented:
Alright, this is looking good, if @jcrist has any more comments, you can get them in another PR, but this is a good point to build off.
